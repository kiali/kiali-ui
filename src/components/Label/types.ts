export interface Labels {
  name: string;
  value: string;
}
