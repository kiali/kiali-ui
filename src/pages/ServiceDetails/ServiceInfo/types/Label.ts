export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
