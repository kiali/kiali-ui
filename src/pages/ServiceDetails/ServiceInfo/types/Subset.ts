import { TrafficPolicy } from '../../../../types/IstioObjects';
import Label from './Label';

export default class Subset {
  name: string;
  labels: { [key: string]: string };
  trafficPolicy: TrafficPolicy;

  constructor(name: string, labels: { [key: string]: string }, trafficPolicy: TrafficPolicy) {
    this.name = name;
    this.labels = labels;
    this.trafficPolicy = trafficPolicy;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels() && this.hasValidTrafficPolicy();
  }

  hasValidName() {
    return typeof this.name === 'string';
  }

  hasValidLabels() {
    const valid = Object.keys(this.labels).every((k, i) => new Label(k, this.labels[k]).isValid());
    return this.labels instanceof Object && valid;
  }

  hasValidTrafficPolicy() {
    return true;
  }
}
