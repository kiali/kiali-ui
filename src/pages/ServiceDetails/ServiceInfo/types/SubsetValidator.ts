import { Subset } from '../../../../types/IstioObjects';

export default class SubsetValidator {
  subset: Subset;

  constructor(subset: Subset) {
    this.subset = subset;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels();
  }

  hasValidName() {
    return this.hasStringType(this.subset.name);
  }

  hasValidLabels() {
    const valid = Object.keys(this.subset.labels).every((k, _i) => this.hasValidLabel(k, this.subset.labels[k]));
    return this.subset.labels instanceof Object && valid;
  }

  hasValidLabel(name: string, value: string): boolean {
    return this.hasStringType(name) && this.hasStringType(value);
  }

  hasStringType(value: any) {
    return typeof value === 'string';
  }
}
