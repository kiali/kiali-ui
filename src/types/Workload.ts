import Namespace from './Namespace';
import { WorkloadHealth } from './Health';
import { Pod, Service } from './IstioObjects';

export interface WorkloadId {
  namespace: string;
  workload: string;
}

export interface Workload {
  name: string;
  type: string;
  createdAt: string;
  resourceVersion: string;
  istioSidecar: boolean;
  labels: { [key: string]: string };
  appLabel: boolean;
  versionLabel: boolean;
  replicas: Number;
  availableReplicas: Number;
  pods: Pod[];
  services: Service[];
  runtimes: Runtime[];
}

export const emptyWorkload: Workload = {
  name: '',
  type: '',
  createdAt: '',
  resourceVersion: '',
  istioSidecar: false,
  labels: {},
  appLabel: false,
  versionLabel: false,
  replicas: 0,
  availableReplicas: 0,
  pods: [],
  services: [],
  runtimes: []
};

export const worloadLink = (ns: string, name: string) => {
  return `/namespaces/${ns}/workloads/${name}`;
};

export const WorkloadIcon = 'bundle';

export const WorkloadType = {
  CronJob: 'CronJob',
  DaemonSet: 'DaemonSet',
  Deployment: 'Deployment',
  DeploymentConfig: 'DeploymentConfig',
  Job: 'Job',
  Pod: 'Pod',
  ReplicaSet: 'ReplicaSet',
  ReplicationController: 'ReplicationController',
  StatefulSet: 'StatefulSet'
};

export interface WorkloadOverview {
  name: string;
  type: string;
  istioSidecar: boolean;
  appLabel: boolean;
  versionLabel: boolean;
}

export interface WorkloadListItem {
  namespace: string;
  workload: WorkloadOverview;
  healthPromise: Promise<WorkloadHealth>;
}

export interface WorkloadNamespaceResponse {
  namespace: Namespace;
  workloads: WorkloadOverview[];
}

export interface Runtime {
  name: string;
  dashboardRefs: DashboardRef[];
}

export interface DashboardRef {
  template: string;
  title: string;
}
