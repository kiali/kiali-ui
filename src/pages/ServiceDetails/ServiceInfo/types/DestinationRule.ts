import { DestinationRuleSpec, K8sMetadata } from '../../../../types/IstioObjects';
import Subset from './Subset';

export default class DestinationRule {
  metadata: K8sMetadata;
  spec: DestinationRuleSpec;
  unformattedField: string;

  constructor(metadata: K8sMetadata, spec: DestinationRuleSpec) {
    this.metadata = metadata;
    this.spec = spec;
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.spec !== null;
  }

  hasValidSubsets() {
    if (!this.spec.subsets) {
      return true;
    }

    let valid = this.spec.subsets instanceof Array;
    valid =
      valid &&
      this.spec.subsets.every((subset, i, ary) =>
        new Subset(subset.name, subset.labels, subset.trafficPolicy).isValid()
      );

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
