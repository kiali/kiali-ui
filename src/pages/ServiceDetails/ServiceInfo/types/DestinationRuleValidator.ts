import { DestinationRule } from '../../../../types/IstioObjects';
import SubsetValidator from './SubsetValidator';

export default class DestinationRuleValidator {
  destinationRule: DestinationRule;
  unformattedField: string;

  constructor(destinationRule: DestinationRule) {
    this.destinationRule = destinationRule;
    this.unformattedField = 'none';
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.destinationRule.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.destinationRule.spec !== null;
  }

  hasValidSubsets() {
    if (!this.destinationRule.spec.subsets) {
      return true;
    }

    let valid = this.destinationRule.spec.subsets instanceof Array;
    valid =
      valid && this.destinationRule.spec.subsets.every((subset, _i, _ary) => new SubsetValidator(subset).isValid());

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.destinationRule.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
