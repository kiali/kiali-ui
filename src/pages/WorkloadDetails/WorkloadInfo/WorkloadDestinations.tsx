import * as React from 'react';
import { Link } from 'react-router-dom';
import { DestinationService } from '../../../types/Workload';

interface WorkloadDestinationsProps {
  destinationServices: DestinationService[];
}

class WorkloadDestinations extends React.Component<WorkloadDestinationsProps> {
  render() {
    return (
      <ul className="workload-destination-services">
        {this.props.destinationServices.map(service => (
          <li key={`dest-service-${service.name}`}>
            <Link to={`/namespaces/${service.namespace}/services/${service.name}`}>{service.name}</Link>
          </li>
        ))}
      </ul>
    );
  }
}

export default WorkloadDestinations;
