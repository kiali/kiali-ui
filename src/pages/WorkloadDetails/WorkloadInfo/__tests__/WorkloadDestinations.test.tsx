import * as React from 'react';
import { render } from 'enzyme';
import WorkloadDestinations from '../WorkloadDestinations';
import { BrowserRouter as Router } from 'react-router-dom';
import { DestinationService } from '../../../../types/Workload';

describe('#WorkloadDestinations renders correctly', () => {
  it('renders a list of links to service details page', () => {
    const services: DestinationService[] = [
      { name: 'reviews', namespace: 'bookinfo' },
      { name: 'istio-citadel', namespace: 'istio-system' }
    ];

    const wrapper = render(
      <Router>
        <div>
          <WorkloadDestinations workloadName={'product-v1'} destinationServices={services} />
        </div>
      </Router>
    );

    const expectationData = {
      reviews: { href: '/namespaces/bookinfo/services/reviews' },
      'istio-citadel': { href: '/namespaces/istio-system/services/istio-citadel' }
    };

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();

    wrapper.find('a').each((index, selector) => {
      const serviceName = selector.children[0].data;
      expect(serviceName).toBeDefined();
      if (!serviceName) {
        return;
      }

      expect(selector.attribs['href']).toEqual(expectationData[serviceName]['href']);
    });
  });
});
