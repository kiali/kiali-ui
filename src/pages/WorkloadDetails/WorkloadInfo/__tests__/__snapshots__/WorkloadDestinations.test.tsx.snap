// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#WorkloadDestinations renders correctly renders a list of links to service details page 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "workload-destination-services",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {
                  "href": "/namespaces/bookinfo/services/reviews",
                },
                "children": Array [
                  Object {
                    "data": "reviews",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "a",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "href": undefined,
                },
                "x-attribsPrefix": Object {
                  "href": undefined,
                },
              },
            ],
            "name": "li",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {
                    "href": "/namespaces/istio-system/services/istio-citadel",
                  },
                  "children": Array [
                    Object {
                      "data": "istio-citadel",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "a",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "href": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "href": undefined,
                  },
                },
              ],
              "name": "li",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {
                  "href": "/namespaces/istio-system/services/istio-citadel",
                },
                "children": Array [
                  Object {
                    "data": "istio-citadel",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "a",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "href": undefined,
                },
                "x-attribsPrefix": Object {
                  "href": undefined,
                },
              },
            ],
            "name": "li",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {
                    "href": "/namespaces/bookinfo/services/reviews",
                  },
                  "children": Array [
                    Object {
                      "data": "reviews",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "a",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "href": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "href": undefined,
                  },
                },
              ],
              "name": "li",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "ul",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
